testutils_test.go

```diff
@@ -114,13 +121,189 @@ func expectTraverse(t *testing.T, trav traverser, start, end string, count int)
+}
+
+func getRandomizedTreeAndMirror(t *testing.T) (*MutableTree, map[string]string) {
+	const cacheSize = 100
+
+	tree, err := getTestTree(cacheSize)
+	require.NoError(t, err)
+
+	mirror := make(map[string]string)
+
+	randomizeTreeAndMirror(t, tree, mirror)
+	return tree, mirror
+}
+
+func randomizeTreeAndMirror(t *testing.T, tree *MutableTree, mirror map[string]string) {
+	if mirror == nil {
+		mirror = make(map[string]string)
+	}
+	const keyValLength = 5
+
+	numberOfSets := 1000
+	numberOfUpdates := numberOfSets / 4
+	numberOfRemovals := numberOfSets / 4
+
+	for numberOfSets > numberOfRemovals*3 {
+		key := randBytes(keyValLength)
+		value := randBytes(keyValLength)
+
+		isUpdated := tree.Set(key, value)
+		require.False(t, isUpdated)
+		mirror[string(key)] = string(value)
+
+		numberOfSets--
+	}
+
+	for numberOfSets+numberOfRemovals+numberOfUpdates > 0 {
+		randOp := rand.Intn(2)
+		if randOp == 0 && numberOfSets > 0 {
+
+			numberOfSets--
+
+			key := randBytes(keyValLength)
+			value := randBytes(keyValLength)
+
+			isUpdated := tree.Set(key, value)
+			require.False(t, isUpdated)
+			mirror[string(key)] = string(value)
+		} else if randOp == 1 && numberOfUpdates > 0 {
+
+			numberOfUpdates--
+
+			key := getRandomKeyFrom(mirror)
+			value := randBytes(keyValLength)
+
+			isUpdated := tree.Set([]byte(key), value)
+			require.True(t, isUpdated)
+			mirror[string(key)] = string(value)
+		} else if numberOfRemovals > 0 {
+
+			numberOfRemovals--
+
+			key := getRandomKeyFrom(mirror)
+
+			val, isRemoved := tree.Remove([]byte(key))
+			require.True(t, isRemoved)
+			require.NotNil(t, val)
+			delete(mirror, string(key))
+		}
+	}
+}
+
+func getRandomKeyFrom(mirror map[string]string) string {
+	keys := make([]string, 0, len(mirror))
+	for k := range mirror {
+		keys = append(keys, k)
+	}
+	key := keys[rand.Intn(len(keys))]
+	return key
+}
+
+func setupMirrorForIterator(t *testing.T, config *iteratorTestConfig, tree *MutableTree) [][]string {
+	mirror := make([][]string, 0)
+
+	startByteToSet := config.startByteToSet
+	endByteToSet := config.endByteToSet
+
+	if !config.ascending {
+		startByteToSet, endByteToSet = endByteToSet, startByteToSet
+	}
+
+	curByte := startByteToSet
+	for curByte != endByteToSet {
+		value := randBytes(5)
+
+		if (config.startIterate == nil || curByte >= config.startIterate[0]) && (config.endIterate == nil || curByte < config.endIterate[0]) {
+			mirror = append(mirror, []string{string(curByte), string(value)})
+		}
+
+		isUpdated := tree.Set([]byte{curByte}, value)
+		require.False(t, isUpdated)
+
+		if config.ascending {
+			curByte++
+		} else {
+			curByte--
+		}
+	}
+	_, _, err := tree.SaveVersion()
+	require.NoError(t, err)
+	return mirror
+}
+
+func assertIterator(t *testing.T, itr dbm.Iterator, mirror [][]string, ascending bool) {
+	startIdx, endIdx := 0, len(mirror)-1
+	increment := 1
+	if !ascending {
+		startIdx, endIdx = endIdx, startIdx
+		increment *= -1
+	}
+
+	for startIdx < endIdx {
```

Comment:
How does this work when !ascending, doesn't startIdX = len(mirror) - 1, and endIdX = 0 ?

Reply:
This was completely wrong. Thanks for catching that. Fixed now